lib/puppet/pops/lookup.rb

```diff
@@ -19,7 +21,7 @@ module Lookup
   # @param lookup_invocation [Invocation] Invocation data containing scope, overrides, and defaults
   # @return [Object] The found value
   #
-  def self.lookup(name, value_type, default_value, has_default, merge, lookup_invocation)
+  def self.lookup(name, value_type, default_value, has_default, merge, convert_to, lookup_invocation)
```

Comment:
This change of the global and public lookup API is very unfortunate. I don't think it's unlikely that this method is actually used by third-party software.

Is it really needed? Can't conversion be done by the caller instead of passing it through here? Or do we expect values being merged by lookup after they have been converted?

Reply:
If it is not done here, the work needs to be done in a context where some of the information is not available. It also needs to be done multiple times for the lookup function, for APL, and for the CLI.

The expansion into individual arguments may be fine, but it is unfortunate that it is hard to evolve - named args in a newer Ruby would be preferable when there are this many args.

Reply:
Isn't it true that the most prominent case for `convert_to` is APL and since APL cannot give explicit options to lookup, the only possible usage there is use of `lookup_options`? Assuming the answer is yes, then the "multiple times" is down to two. One might even argue that it doesn't make much sense for in the lookup CLI either since it will be converted to a printable form anyway. Some corner cases perhaps, but still.

Given that this is public API and adding another option must be considered a breakage (and hence move this into 6.0), wouldn't it be better to just skip this extra option?